File: game_engine/hints.py
from typing import List
from game_engine.board import Board
from game_engine.models import GameEventResult
from game_engine.validation import validate_move
from game_engine.message_codes import CAPTURE_CONTINUE_SAME
from game_engine.dictionaries import (
    shatra_and_biy_possible_captures,
    batyr_moves_and_captures,
    black_shatra_possible_moves,
    white_shatra_possible_moves,
    black_biy_possible_moves,
    white_biy_possible_moves,
)


def _get_all_candidates(
    cells: dict,
    current_color: str,
    from_cell: int,
    piece_type: str
) -> List[int]:
    """Собирает все возможные цели из словарей (ходы + взятия)."""
    candidates = set()

    if piece_type in ["шатра", "бий"]:
        moves = None
        if piece_type == "шатра":
            moves = black_shatra_possible_moves if current_color == "черный" else white_shatra_possible_moves
        else:
            moves = black_biy_possible_moves if current_color == "черный" else white_biy_possible_moves
        for target in moves.get(from_cell, []):
            candidates.add(target)
        for target in shatra_and_biy_possible_captures.get(from_cell, {}):
            candidates.add(target)

    elif piece_type == "батыр":
        for direction in batyr_moves_and_captures.get(from_cell, []):
            for target in direction:
                candidates.add(target)

    return list(candidates)


def get_hints(
    cells: dict,
    current_color: str,
    from_cell: int,
    batyr_captured_this_turn: List[int] = None,
    chain_capture_cell: int = None
) -> GameEventResult:
    """Возвращает подсказки (куда можно пойти) для фигуры на from_cell.
    
    Простая прослойка: собирает кандидатов из словарей, прогоняет через
    validate_move(), подсвечивает то, что вернуло True.
    """
    if batyr_captured_this_turn is None:
        batyr_captured_this_turn = []

    board = Board(cells)
    piece = board.get_piece_object(from_cell)
    if not piece or piece.get_color() != current_color:
        return GameEventResult(essential_positions=[], captured_pieces=batyr_captured_this_turn.copy())

    # Если есть цепочка — показываем только взятия этой фигурой
    if chain_capture_cell and chain_capture_cell != 0:
        if from_cell != chain_capture_cell:
            return GameEventResult(
                essential_positions=[],
                captured_pieces=batyr_captured_this_turn.copy(),
                message_code=CAPTURE_CONTINUE_SAME,
            )
        return _get_chain_hints(cells, current_color, from_cell, batyr_captured_this_turn, piece)

    # Собираем всех кандидатов и прогоняем через validate_move
    candidates = _get_all_candidates(cells, current_color, from_cell, piece.get_type())
    allowed = []
    for target in candidates:
        valid, _ = validate_move(
            cells, from_cell, target, current_color,
            batyr_captured_this_turn, check_mandatory=True
        )
        if valid:
            allowed.append(target)

    return GameEventResult(
        essential_positions=allowed,
        captured_pieces=batyr_captured_this_turn.copy()
    )


def _get_chain_hints(
    cells: dict,
    current_color: str,
    from_cell: int,
    batyr_captured_this_turn: List[int],
    piece
) -> GameEventResult:
    """Подсказки для продолжения цепочки взятий."""
    allowed = []
    if piece.get_type() in ["шатра", "бий"]:
        for to_cell in shatra_and_biy_possible_captures.get(from_cell, {}):
            if piece.can_capture(cells, from_cell, to_cell, batyr_captured_this_turn):
                allowed.append(to_cell)
    elif piece.get_type() == "батыр":
        for direction in batyr_moves_and_captures.get(from_cell, []):
            for to_cell in direction:
                if piece.can_capture(cells, from_cell, to_cell, batyr_captured_this_turn):
                    allowed.append(to_cell)
    return GameEventResult(
        essential_positions=allowed,
        captured_pieces=batyr_captured_this_turn.copy()
    )

File: frontend/packages/shatra-rules/src/hints.js
import { Board } from './board.js';
import {
  shatraAndBiyPossibleCaptures,
  batyrMovesAndCaptures,
  blackShatraPossibleMoves,
  whiteShatraPossibleMoves,
  blackBiyPossibleMoves,
  whiteBiyPossibleMoves,
} from './dictLoader.js';

export function getAllMandatoryCaptures(board, color, batyrCapturedThisTurn = null) {
  const caps = batyrCapturedThisTurn || [];
  const mandatory = [];
  const cells = board.cells;

  for (const [pos, piece] of board.getAllPieces()) {
    if (piece.getColor() !== color) continue;

    const pieceType = piece.getType();
    const candidates = new Set();

    if (pieceType === 'шатра' || pieceType === 'бий') {
      for (const toCell of Object.keys(shatraAndBiyPossibleCaptures[pos] || {})) {
        candidates.add(Number(toCell));
      }
    } else if (pieceType === 'батыр') {
      for (const direction of batyrMovesAndCaptures[pos] || []) {
        for (const target of direction) candidates.add(target);
      }
    }

    for (const toCell of [...candidates].sort((a, b) => a - b)) {
      if (piece.canCapture(cells, pos, toCell, caps)) {
        mandatory.push([pos, toCell]);
      }
    }
  }

  return mandatory;
}

export function validateMove(
  cells,
  fromCell,
  toCell,
  currentColor,
  batyrCapturedThisTurn = null,
  checkMandatory = true,
  chainCaptureCell = null,
) {
  const caps = batyrCapturedThisTurn || [];
  const board = new Board(cells);

  const piece = board.getPieceObject(fromCell);
  if (!piece) return { valid: false, code: 'NO_PIECE' };
  if (piece.getColor() !== currentColor) return { valid: false, code: 'WRONG_COLOR' };
  if (cells[toCell] != null) return { valid: false, code: 'TARGET_OCCUPIED' };

  if (checkMandatory) {
    const mandatoryCaptures = getAllMandatoryCaptures(board, currentColor, caps);
    if (mandatoryCaptures.length > 0) {
      const hasNonBiyAttacker = mandatoryCaptures.some(([f]) => {
        const p = board.getPieceObject(f);
        return p && p.getType() !== 'бий';
      });
      const attackerPositions = new Set(mandatoryCaptures.map(([f]) => f));

      if (!attackerPositions.has(fromCell)) {
        if (hasNonBiyAttacker) {
          return { valid: false, code: 'MANDATORY_CAPTURE_OTHER_PIECE' };
        }
        if (piece.getType() !== 'бий') {
          return { valid: false, code: 'ONLY_BIY_CAN_CAPTURE' };
        }
      } else {
        const captureTargets = new Set(
          mandatoryCaptures.filter(([f]) => f === fromCell).map(([, t]) => t),
        );
        if (!captureTargets.has(toCell)) {
          if (!(piece.getType() === 'бий' && !hasNonBiyAttacker)) {
            return { valid: false, code: 'MANDATORY_CAPTURE_THIS_PIECE' };
          }
        }
      }
    }
  }

  const piece2 = board.getPieceObject(fromCell);
  if (!piece2) return { valid: false, code: 'INTERNAL_NO_PIECE' };

  if (piece2.getType() === 'батыр') {
    for (const direction of batyrMovesAndCaptures[fromCell] || []) {
      if (!direction.includes(toCell)) continue;
      for (const cell of direction) {
        if (cell === toCell) break;
        const cellPiece = cells[cell];
        if (cellPiece && cellPiece.includes(currentColor)) {
          return { valid: false, code: 'OWN_PIECE_BLOCKS_BATYR' };
        }
      }
    }
  }

  if (piece2.canCapture(cells, fromCell, toCell, caps)) {
    return { valid: true, code: 'OK_CAPTURE' };
  }
  if (piece2.canMove(cells, fromCell, toCell)) {
    return { valid: true, code: 'OK_MOVE' };
  }

  return { valid: false, code: 'ILLEGAL_MOVE' };
}

export function findCapturedEnemy(
  cells,
  piece,
  fromCell,
  toCell,
  batyrCapturedThisTurn = null,
) {
  const caps = batyrCapturedThisTurn || [];

  if (piece.getType() === 'шатра' || piece.getType() === 'бий') {
    const captures = shatraAndBiyPossibleCaptures[fromCell] || {};
    return captures[toCell] ?? null;
  }

  if (piece.getType() === 'батыр') {
    const opponentPrefix = piece.getColor() === 'белый' ? 'чер' : 'бел';
    for (const direction of batyrMovesAndCaptures[fromCell] || []) {
      if (!direction.includes(toCell)) continue;
      for (const pos of direction) {
        if (pos === toCell) return null;
        const cellContent = cells[pos];
        if (cellContent && cellContent.includes(opponentPrefix)) {
          if (!caps.includes(pos)) return pos;
          continue;
        }
        if (cellContent != null) return null;
      }
    }
  }

  return null;
}

export function batyrCanContinueCapture(
  board,
  fromCell,
  color,
  batyrCapturedThisTurn = null,
) {
  const caps = batyrCapturedThisTurn || [];
  const piece = board.getPieceObject(fromCell);
  if (!piece || piece.getType() !== 'батыр') return false;

  const cells = board.cells;
  for (const [start, target] of getAllMandatoryCaptures(board, color, caps)) {
    if (start === fromCell && piece.canCapture(cells, fromCell, target, caps)) {
      return true;
    }
  }
  return false;
}

function getAllCandidates(cells, currentColor, fromCell, pieceType) {
  const candidates = new Set();

  if (pieceType === 'шатра' || pieceType === 'бий') {
    const moves = pieceType === 'шатра'
      ? (currentColor === 'черный' ? blackShatraPossibleMoves : whiteShatraPossibleMoves)
      : (currentColor === 'черный' ? blackBiyPossibleMoves : whiteBiyPossibleMoves);
    for (const target of moves[fromCell] || []) candidates.add(target);
    for (const target of Object.keys(shatraAndBiyPossibleCaptures[fromCell] || {})) {
      candidates.add(Number(target));
    }
  } else if (pieceType === 'батыр') {
    for (const direction of batyrMovesAndCaptures[fromCell] || []) {
      for (const target of direction) candidates.add(target);
    }
  }

  return [...candidates];
}

function collectCaptureHighlightCells(cells, fromCell, essentialPositions, piece, batyrCapturedThisTurn) {
  if (!piece || !essentialPositions.length) return [];
  const enemies = new Set();
  for (const toCell of essentialPositions) {
    const enemy = findCapturedEnemy(cells, piece, fromCell, toCell, batyrCapturedThisTurn);
    if (enemy != null) enemies.add(enemy);
  }
  return [...enemies].sort((a, b) => a - b);
}

function getChainHints(cells, currentColor, fromCell, batyrCapturedThisTurn, piece) {
  const allowed = [];
  if (piece.getType() === 'шатра' || piece.getType() === 'бий') {
    for (const toCell of Object.keys(shatraAndBiyPossibleCaptures[fromCell] || {})) {
      const t = Number(toCell);
      if (piece.canCapture(cells, fromCell, t, batyrCapturedThisTurn)) {
        allowed.push(t);
      }
    }
  } else if (piece.getType() === 'батыр') {
    for (const direction of batyrMovesAndCaptures[fromCell] || []) {
      for (const toCell of direction) {
        if (piece.canCapture(cells, fromCell, toCell, batyrCapturedThisTurn)) {
          allowed.push(toCell);
        }
      }
    }
  }
  return allowed;
}

/**
 * @param {object} params
 * @param {Record<number, string|null>} params.cells
 * @param {string} params.currentColor
 * @param {number} params.fromCell
 * @param {number[]} [params.batyrCapturedThisTurn]
 * @param {number|null} [params.chainCaptureCell]
 */
export function getHints({
  cells,
  currentColor,
  fromCell,
  batyrCapturedThisTurn = [],
  chainCaptureCell = null,
}) {
  const caps = [...(batyrCapturedThisTurn || [])];
  const board = new Board(cells);
  const piece = board.getPieceObject(fromCell);

  if (!piece || piece.getColor() !== currentColor) {
    return { essentialPositions: [], capturedPieces: caps, captureHighlightCells: [], messageCode: '' };
  }

  if (chainCaptureCell && chainCaptureCell !== 0) {
    if (fromCell !== chainCaptureCell) {
      return {
        essentialPositions: [],
        capturedPieces: caps,
        captureHighlightCells: [],
        messageCode: 'capture.continue_same',
      };
    }
    const chainAllowed = getChainHints(cells, currentColor, fromCell, caps, piece);
    return {
      essentialPositions: chainAllowed,
      capturedPieces: caps,
      captureHighlightCells: collectCaptureHighlightCells(
        cells, fromCell, chainAllowed, piece, caps,
      ),
      messageCode: '',
    };
  }

  const candidates = getAllCandidates(cells, currentColor, fromCell, piece.getType());
  const allowed = [];
  for (const target of candidates) {
    const { valid } = validateMove(cells, fromCell, target, currentColor, caps, true);
    if (valid) allowed.push(target);
  }

  return {
    essentialPositions: allowed,
    capturedPieces: caps,
    captureHighlightCells: collectCaptureHighlightCells(
      cells, fromCell, allowed, piece, caps,
    ),
    messageCode: '',
  };
}
